### ISSUE TITLE
Calculating Gradients for Nested Models

### ISSUE THREAD

**Issue by totti0223** [keras-team/keras#11767]
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
`pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps`

- [x] Check that your version of TensorFlow is up-to-date. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

When I try to construct a model connecting (multiple) model by Functional API, I cannot access to the layers inside the nested models. More accurately, I can, but seems to be not in the parent graph(?). suppose for the bellow connected model that contains model_1 and model_2 inside, K.gradients cannotbe calculated by the two.
```
connected.get_layer("model_2").get_layer("target_dense_1").output
connected. input
```

the first seems to target the layer of the model_2, not the layer of the connected model. I want to target the layer that is inside the connected model for calculating gradients outside the nested model (case1 bellow). Are there any workarounds?

below are codes. the model structure means nothing.
```  python:codes.py
x1_in = layers.Input((7,7,512))
x1 = layers.Flatten()(x1_in)
x1 = layers.Dense(1000)(x1)
x1_out = layers.Dense(1000)(x1)
model1 = Model(inputs=x1_in,outputs=x1_out,name="model_1")
print("model1 summary")
model1.summary()

x2_in = layers.Input((1000,))
x2 = layers.Dense(10,name="target_dense_1")(x2_in)
x2_out = layers.Dense(10)(x2)
model2 = Model(inputs=x2_in,outputs=x2_out,name="model_2")
print("model2 summary")
model2.summary()

inputs = layers.Input((7,7,512))
x = model1(inputs)
outputs = model2(x)
connected = Model(inputs=inputs,outputs=outputs)
print("connected model summary")
connected.summary()

#case 1. calculate the graidents between model2 intermediate output and connected model input
loss = K.mean(connected.get_layer("model_2").get_layer("target_dense_1").output)
grads = K.gradients(loss, connected.input)[0]
print(loss)
print(grads) #returns none

#case2. gradient calculation within the same nested model works
loss = K.mean(connected.get_layer("model_2").get_layer("target_dense_1").output)
grads = K.gradients(loss, connected.get_layer("model_2").inputs[0])[0]
print(loss)
print(grads) #returns shape ?,1000

#case3. the below also works, so this loss is targeting to model_2 itself, not the model_2 in concatenate model
loss = K.mean(connected.get_layer("model_2").get_layer("target_dense_1").output)
grads = K.gradients(loss, model2.inputs[0])[0]
print(grads) #returns shape ?,1000

```
Returns

``` python:Result
model1 summary
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 7, 7, 512)         0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 25088)             0         
_________________________________________________________________
dense_1 (Dense)              (None, 1000)              25089000  
_________________________________________________________________
dense_2 (Dense)              (None, 1000)              1001000   
=================================================================
Total params: 26,090,000
Trainable params: 26,090,000
Non-trainable params: 0
_________________________________________________________________
model2 summary
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         (None, 1000)              0         
_________________________________________________________________
target_dense_1 (Dense)       (None, 10)                10010     
_________________________________________________________________
dense_3 (Dense)              (None, 10)                110       
=================================================================
Total params: 10,120
Trainable params: 10,120
Non-trainable params: 0
_________________________________________________________________
connected model summary
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_3 (InputLayer)         (None, 7, 7, 512)         0         
_________________________________________________________________
model_1 (Model)              (None, 1000)              26090000  
_________________________________________________________________
model_2 (Model)              (None, 10)                10120     
=================================================================
Total params: 26,100,120
Trainable params: 26,100,120
Non-trainable params: 0
_________________________________________________________________
Tensor("Mean:0", shape=(), dtype=float32)
None
Tensor("Mean_1:0", shape=(), dtype=float32)
Tensor("gradients_1/target_dense_1/MatMul_grad/MatMul:0", shape=(?, 1000), dtype=float32)
Tensor("Mean_2:0", shape=(), dtype=float32)
Tensor("gradients_2/target_dense_1/MatMul_grad/MatMul:0", shape=(?, 1000), dtype=float32)
```

**Comment by ghylander** [comment]
any updates on this? it's been more than 2 years now
if you are using transfer learning, downloading an existing model and training it on new data, the base model becomes a nested model
correct visualization of nested models filters/activations/grad-cam is pretty much impossible right now
